aggregator/src/core.rs

```diff
@@ -62,3 +84,680 @@ pub(crate) fn extract_accumulators_and_proof(
+    // each part of the preimage is mapped to image by Keccak256
+    let witness = multi_keccak(preimages, challenges, keccak_round_capacity(num_rows))
+        .map_err(|e| Error::AssertionFailure(format!("multi keccak assignment failed: {e:?}")))?;
+    end_timer!(timer);
+
+    // extract the indices of the rows for which the preimage and the digest cells lie in
+    let (preimage_indices, digest_indices) = get_indices(preimages);
+
+    let (hash_input_cells, hash_output_cells, data_rlc_cells, hash_input_len_cells) = layouter
+        .assign_region(
+            || "assign keccak rows",
+            |mut region| -> Result<
+                (
+                    Vec<AssignedCell<Fr, Fr>>,
+                    Vec<AssignedCell<Fr, Fr>>,
+                    Vec<AssignedCell<Fr, Fr>>,
+                    Vec<AssignedCell<Fr, Fr>>,
+                ),
+                halo2_proofs::plonk::Error,
+            > {
+                #[cfg(feature = "skip_first_pass")]
+                if is_first_time {
+                    is_first_time = false;
+                    let offset = witness.len() - 1;
+                    keccak_config.set_row(&mut region, offset, &witness[offset])?;
+                    return Ok((vec![], vec![], vec![], vec![]));
+                }
+
+                let mut preimage_indices_iter = preimage_indices.iter();
+                let mut digest_indices_iter = digest_indices.iter();
+
+                let mut cur_preimage_index = preimage_indices_iter.next();
+                let mut cur_digest_index = digest_indices_iter.next();
+
+                // ====================================================
+                // Step 1. Extract the hash cells
+                // ====================================================
+                let mut hash_input_cells = vec![];
+                let mut hash_output_cells = vec![];
+                let mut data_rlc_cells = vec![];
+                let mut hash_input_len_cells = vec![];
+
+                let timer = start_timer!(|| "assign row");
+                log::trace!("witness length: {}", witness.len());
+                for (offset, keccak_row) in witness.iter().enumerate() {
+                    let row = keccak_config.set_row(&mut region, offset, keccak_row)?;
+
+                    if cur_preimage_index.is_some() && *cur_preimage_index.unwrap() == offset {
+                        // 10-th column is Keccak input in Keccak circuit
+                        hash_input_cells.push(row[10].clone());
+                        cur_preimage_index = preimage_indices_iter.next();
+                    }
+                    if cur_digest_index.is_some() && *cur_digest_index.unwrap() == offset {
+                        // last column is Keccak output in Keccak circuit
+                        hash_output_cells.push(row.last().unwrap().clone()); // sage unwrap
+                        cur_digest_index = digest_indices_iter.next();
+                    }
```

Comment:
https://github.com/scroll-tech/zkevm-circuits/blob/236b3cdc3d621ab89ee5e46fad23928375055d2d/aggregator/src/aggregation/config.rs#L86-L87

Reply:
instead of hard-code, we can use the params with only their names in Keccak config and build formulae for hash input output rows. I had some formulae in my keccak doc https://hackmd.io/@dieGzUCgSGmRZFQ7SDxXCA/H1vMF0_u2

Reply:
pi circuit has a design feature of directly extracting needed cells into an array and do copy constraints directly.